src/test/java/com/google/sps/AuthenticationTest.java

```diff
@@ -0,0 +1,54 @@
+package com.google.sps;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mockito;
+import static org.mockito.Mockito.when;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.users.User;
+import java.util.*;
+
+@RunWith(JUnit4.class)
+public final class AuthenticationTest {
+
+    public String testAuth(UserService mock){
+        if(mock.isUserLoggedIn()){
+            String userEmail = mock.getCurrentUser().getEmail();
+            String urlToRedirectToAfterUserLogsOut = "/";
+            String logoutUrl = mock.createLogoutURL(urlToRedirectToAfterUserLogsOut);
+            return userEmail;  
+        }
+        else{
+            String urlToRedirectToAfterUserLogsIn = "redirect";
+            String loginUrl = mock.createLoginURL(urlToRedirectToAfterUserLogsIn);
+            return "<center><a href=\"" + loginUrl + "\"><button style=\"background-color: cyan; border-radius:12px;\"><span style=\"color:white;\">Login</span></button/></a></center>";
+        }
+    }
```

Comment:
This function isn't testing the actual AuthenticationServlet. Please call the servlet directly instead of having it's code in here.

Reply:
+1. Move original servlet logic to a helper function and call this helper in the servlet and here.